service/storage_service.cc

```diff
@@ -2658,20 +2658,207 @@ void storage_service::add_expire_time_if_found(inet_address endpoint, int64_t ex
 
-// For more details, see the commends on column_family::load_new_sstables
-// All the global operations are going to happen here, and just the reloading happens
-// in there.
-future<> storage_service::load_new_sstables(sstring ks_name, sstring cf_name) {
-    if (_loading_new_sstables) {
-        throw std::runtime_error("Already loading SSTables. Try again later");
-    } else {
-        _loading_new_sstables = true;
+class send_meta_data {
+    gms::inet_address _node;
+    rpc::sink<frozen_mutation_fragment, streaming::stream_mutation_fragments_cmd> _sink;
+    rpc::source<int32_t> _source;
+    bool _error_from_peer = false;
+    size_t _num_partitions_sent = 0;
+    size_t _num_bytes_sent = 0;
+public:
+    send_meta_data(gms::inet_address node,
+            rpc::sink<frozen_mutation_fragment, streaming::stream_mutation_fragments_cmd> sink,
+            rpc::source<int32_t> source)
+        : _node(std::move(node))
+        , _sink(std::move(sink))
+        , _source(std::move(source)) {
+    }
+    void receive() {
+        for (;;) {
+            std::optional<std::tuple<int32_t>> status_opt = _source().get0();
+            if (!status_opt) {
+                slogger.debug("Finished reading source from node={}", _node);
+                break;
+            }
+            auto status = std::get<0>(*status_opt);
+            slogger.debug("Got error code from node={}, status={}", _node, status);
+            if (status == -1) {
+                _error_from_peer = true;
+            }
+        }
+    }
+    future<> send(const frozen_mutation_fragment& fmf, bool is_partition_start) {
+        auto size = fmf.representation().size();
+        if (is_partition_start) {
+            ++_num_partitions_sent;
+        }
+        _num_bytes_sent += size;
+        slogger.trace("Send mf to node={}, size={}", _node, size);
+        return _sink(fmf, streaming::stream_mutation_fragments_cmd::mutation_fragment_data);
+    }
+    void error() {
+        slogger.debug("Send error to node={}", _node);
+        _sink(frozen_mutation_fragment(bytes_ostream()), streaming::stream_mutation_fragments_cmd::error).get();
+        slogger.debug("Closed sink to node={}", _node);
+        _sink.close().get();
+    }
+    void finish() {
+        slogger.debug("Send eos to node={}", _node);
+        _sink(frozen_mutation_fragment(bytes_ostream()), streaming::stream_mutation_fragments_cmd::end_of_stream).get();
+        receive();
+        slogger.debug("Closed sink to node={}", _node);
+        _sink.close().get();
     }
+    size_t num_partitions_sent() {
+        return _num_partitions_sent;
+    }
+    size_t num_bytes_sent() {
+        return _num_bytes_sent;
+    }
+};
+
+future<> storage_service::load_and_stream(sstring ks_name, sstring cf_name,
+        utils::UUID table_id, std::vector<sstables::shared_sstable> sstables, bool primary_replica_only) {
+    seastar::thread_attributes attr;
+    attr.sched_group = _db.local().get_streaming_scheduling_group();
```

Comment:
Again not needed. And you can use a coroutine.

Reply:
Here, I actually want to make it explicit. Because we may want to run it under a different group other than streaming. For example, repair and load_and_stream run in parallel. We may want to make load_and_stream have even lower priority.

Reply:
storage_service doesn't deal with scheduling groups now. I'd like to keep it this way and not spread out the information about what runs where all over the source base. Right now it is inheritance based and main makes most of the decisions.

Reply:
Ok. removed all attr.